tests/all/cli_tests.rs

```diff
@@ -1548,61 +1549,73 @@ mod test_programs {
-            }) {
-                Some(addr) => {
-                    assert!(reader.buffer().is_empty());
-                    child.stderr = Some(reader.into_inner());
-                    Ok(WasmtimeServe {
-                        child: Some(child),
-                        addr,
-                    })
+            let mut read_addr_from_line = |prefix: &str| -> Result<SocketAddr> {
+                reader.read_line(&mut line)?;
+
+                if !line.starts_with(prefix) {
+                    bail!("input line `{line}` didn't start with `{prefix}`");
+                }
+                match line.find("127.0.0.1").and_then(|addr_start| {
+                    let addr = &line[addr_start..];
+                    let addr_end = addr.find("/")?;
+                    addr[..addr_end].parse().ok()
+                }) {
+                    Some(addr) => {
+                        line.truncate(0);
+                        Ok(addr)
+                    }
+                    None => bail!("failed to address from: {line}"),
                 }
-                None => {
+            };
+            let shutdown_addr = read_addr_from_line("Listening for shutdown");
+            let addr = read_addr_from_line("Serving HTTP on");
+            let (shutdown_addr, addr) = match (shutdown_addr, addr) {
+                (Ok(a), Ok(b)) => (a, b),
+                // If either failed kill the child and otherwise try to shepherd
+                // along any contextual information we have.
+                (Err(a), _) | (_, Err(a)) => {
                     child.kill()?;
                     child.wait()?;
                     reader.read_to_string(&mut line)?;
-                    bail!("failed to start child: {line}")
+                    return Err(a.context(line));
                 }
-            }
+            };
+            assert!(reader.buffer().is_empty());
+            child.stderr = Some(reader.into_inner());
+            Ok(WasmtimeServe {
+                child: Some(child),
+                addr,
+                shutdown_addr,
+            })
         }
 
         /// Completes this server gracefully by printing the output on failure.
         fn finish(mut self) -> Result<(String, String)> {
             let mut child = self.child.take().unwrap();
 
-            // If the child process has already exited then collect the output
-            // and test if it succeeded. Otherwise it's still running so kill it
-            // and then reap it. Assume that if it's still running then the test
-            // has otherwise passed so no need to print the output.
-            let known_failure = if child.try_wait()?.is_some() {
-                false
-            } else {
-                child.kill()?;
-                true
-            };
+            // If the child process has already exited, then great! Otherwise
+            // the server is still running so send it a graceful ctrl-c shutdown
```

Comment:
It doesn't look like we send a ctrl-c here, regardless of platform, so maybe change this comment to read "...so connect to the shutdown port..." or something similar.

Reply:
Oops sorry I was halfway through updating this comment when I context switched at some point and forgot to update the other half of the comment...